File: pwa/src/lib/contracts.ts
export type ChartType = 'line' | 'bar' | 'pie' | 'gauge' | 'area' | 'tank' | 'report' | 'kpi' | 'heatmap' | 'waterfall' | 'button' | 'multiblock'

// ── Tank Config ───────────────────────────────────────────────────────────────
export interface TankProductColor {
  produto: string
  color:   string
}

export interface TankConfig {
  field_produto:     string
  field_volume:      string
  field_capacidade:  string
  field_disponivel?: string
  field_percentual?: string
  unidade:           string
  threshold_low:     number
  threshold_mid:     number
  colunas:           1 | 2

  // Aparência
  font_size_produto?:    number
  font_size_volume?:     number
  font_size_percentual?: number
  product_colors?:       TankProductColor[]
}
export type UserRole  = 'operador' | 'gerente' | 'dono'

// ── Report Config ─────────────────────────────────────────────────────────────
export interface ReportColumn {
  field:    string
  label:    string
  format?:  'text' | 'number' | 'currency' | 'percent'
  align?:   'left' | 'right' | 'center'
  summary?: 'none' | 'sum'
}

export interface ReportConfig {
  columns:     ReportColumn[]
  showIndex?:  boolean
  show_index?: boolean
  showTotals?: boolean
  show_totals?: boolean
}

export interface ChartMetadata {
  id: string
  nome: string
  chart_type: ChartType
  query: {
    sql: string
    refresh_seconds: number
    timeout_seconds: number
  }
  axes: {
    x: { field: string; label: string; format?: string }
    y: Array<{ field: string; label: string; format?: string; color?: string }>
  }
  display: {
    height: 'sm' | 'md' | 'lg'
    // Aceita tanto snake_case (config original) quanto camelCase (API com postgres.camel)
    show_legend?:  boolean
    showLegend?:   boolean
    show_tooltip?: boolean
    showTooltip?:  boolean
    gradient?:     boolean
    // Ícone + cores nos cards de menu PWA
    icon?:        string
    icon_color?:  string
    iconColor?:   string
    icon_bg?:     string
    iconBg?:      string
  }
  permissions: {
    min_role?: UserRole  // snake_case (config original)
    minRole?:  UserRole  // camelCase (API com postgres.camel)
  }
  report_config?:     ReportConfig
  tank_config?:       TankConfig
  multiblock_config?: MultiblockConfig
  date_filter?:       TemplateDateFilter
  categoria?:         string
  descricao?:         string
  is_publico?:        boolean
}

// ── Filtro de data ───────────────────────────────────────────────────────────
export type DatePreset = 'today' | 'yesterday' | 'last_7d' | 'last_30d' | 'this_month' | 'last_month' | 'custom'

export interface TemplateDateFilter {
  enabled:         boolean
  filter_type:     'date' | 'datetime'
  param_inicio:    string
  param_fim:       string
  default_preset:  DatePreset
  default_from?:   string
  default_to?:     string
}

// ── Multiblock — 1 SQL → N visualizações ─────────────────────────────────────
export type AggregateFn = 'sum' | 'avg' | 'min' | 'max' | 'count' | 'first' | 'last'
export type KpiFormat   = 'number' | 'currency' | 'percent' | 'litros' | 'text'

export interface MbKpiBlock {
  type:    'kpi'
  title:   string
  field:   string
  agg:     AggregateFn
  format:  KpiFormat
  icon?:   string
  color?:  string
}
export interface MbDonutBlock {
  type:        'donut'
  title:       string
  groupBy:     string
  valueField:  string
  valueAgg?:   AggregateFn
  topN?:       number
  showLegend?: boolean
}
export interface MbTableBlock {
  type:        'table'
  title?:      string
  columns:     ReportColumn[]
  showTotals?: boolean
  maxRows?:    number
}
export interface MbBarBlock {
  type:         'bar'
  title:        string
  xField:       string
  yField:       string
  yAgg?:        AggregateFn
  orientation?: 'h' | 'v'
}
export type MultiBlock = MbKpiBlock | MbDonutBlock | MbTableBlock | MbBarBlock

export interface MultiblockConfig {
  blocks:  MultiBlock[]
  layout?: { preset?: 'kpis-top' | 'side-by-side' | 'stacked' }
}

export interface MQTTCommand {
  type: 'READ_QUERY' | 'WRITE_COMMAND' | 'SYNC_TEMPLATE'
  request_id: string
  template_id?: string
  user_jwt: string
  empresas_ids: number[]
  force_refresh?: boolean
  timestamp: number
}

export interface MQTTWriteCommand extends MQTTCommand {
  type: 'WRITE_COMMAND'
  subtype: 'PRECO_UPDATE' | 'DESCONTO_RESPONSE'
  payload: Record<string, unknown>
}

export interface MQTTResponse {
  type: 'READ_RESULT' | 'WRITE_RESULT' | 'ERROR'
  request_id: string
  template_id?: string
  status: 'success' | 'error' | 'denied'
  data?: Record<string, unknown>[]
  cached?: boolean
  cache_age_seconds?: number
  row_count?: number
  error_message?: string
  timestamp: number
}

export interface DescontoRequest {
  type: 'DESCONTO_REQUEST'
  notification_id: string
  bico: number
  litros: number
  valor_total: number
  desconto_solicitado: number
  operador_id:    string
  operador_nome?: string   // nome legível — preenchido pelo agente Go
  timestamp: number
}

export interface UserSession {
  id: string
  nome: string
  email: string
  role: UserRole
  cliente_id: string
  cnpj: string
  empresas: Array<{ id: number; nome: string; is_master: boolean; codigoErp?: number; codigo_erp?: number }>
  jwt: string
  expires_at: number
}

File: agent/pkg/models/contracts.go
package models

import "time"

// ChartType representa os tipos de gráfico suportados
type ChartType string

const (
	ChartLine  ChartType = "line"
	ChartBar   ChartType = "bar"
	ChartPie   ChartType = "pie"
	ChartGauge ChartType = "gauge"
	ChartArea  ChartType = "area"
)

// UserRole define os níveis de permissão
type UserRole string

const (
	RoleOwner    UserRole = "dono"
	RoleManager  UserRole = "gerente"
	RoleOperator UserRole = "operador"
)

// CommandType tipos de comandos recebidos via MQTT
type CommandType string

const (
	CmdReadQuery        CommandType = "READ_QUERY"
	CmdWriteCmd         CommandType = "WRITE_COMMAND"
	CmdSyncTemplate     CommandType = "SYNC_TEMPLATE"
	CmdUpdateAgent      CommandType = "UPDATE_AGENT"
	CmdDiscoverEmpresas CommandType = "DISCOVER_EMPRESAS"
)

// WriteSubtype subtipos de comandos de escrita
type WriteSubtype string

const (
	SubtypePrecoUpdate   WriteSubtype = "PRECO_UPDATE"
	SubtypeDescontoResp  WriteSubtype = "DESCONTO_RESPONSE"
)

// MQTTCommand é o contrato base de comandos recebidos do MQTT
type MQTTCommand struct {
	Type          CommandType   `json:"type"`
	RequestID     string        `json:"request_id"`
	TemplateID    string        `json:"template_id,omitempty"`
	UserJWT       string        `json:"user_jwt"`
	EmpresasIDs   []int64       `json:"empresas_ids"`
	ForceRefresh  bool          `json:"force_refresh,omitempty"`
	Timestamp     int64         `json:"timestamp"`
	ResponseTopic string        `json:"response_topic,omitempty"`

	// Filtro de período (ISO date "YYYY-MM-DD" ou datetime "YYYY-MM-DD HH:MM:SS")
	// Usado para substituir :data_inicio/:data_fim e :datetime_inicio/:datetime_fim no SQL
	DateFrom string `json:"date_from,omitempty"`
	DateTo   string `json:"date_to,omitempty"`

	// Campos para WRITE_COMMAND
	Subtype WriteSubtype           `json:"subtype,omitempty"`
	Payload map[string]interface{} `json:"payload,omitempty"`
}

// MQTTResponse é o contrato de respostas enviadas ao MQTT
type MQTTResponse struct {
	Type          string                   `json:"type"`
	RequestID     string                   `json:"request_id"`
	TemplateID    string                   `json:"template_id,omitempty"`
	Status        string                   `json:"status"` // success | error | denied
	Data          []map[string]interface{} `json:"data,omitempty"`
	Cached        bool                     `json:"cached"`
	CacheAgeSeconds int64                  `json:"cache_age_seconds,omitempty"`
	RowCount      int                      `json:"row_count,omitempty"`
	ErrorMessage  string                   `json:"error_message,omitempty"`
	Timestamp     int64                    `json:"timestamp"`
}

// ChartMetadata define um template de gráfico
type ChartMetadata struct {
	ID        string    `json:"id"`
	Nome      string    `json:"nome"`
	ChartType ChartType `json:"chart_type"`
	Query     struct {
		SQL            string `json:"sql"`
		RefreshSeconds int    `json:"refresh_seconds"`
		TimeoutSeconds int    `json:"timeout_seconds"`
	} `json:"query"`
	Axes struct {
		X struct {
			Field  string `json:"field"`
			Label  string `json:"label"`
			Format string `json:"format,omitempty"`
		} `json:"x"`
		Y []struct {
			Field  string `json:"field"`
			Label  string `json:"label"`
			Format string `json:"format,omitempty"`
			Color  string `json:"color,omitempty"`
		} `json:"y"`
	} `json:"axes"`
	Display struct {
		Height      string `json:"height"`
		ShowLegend  bool   `json:"show_legend"`
		ShowTooltip bool   `json:"show_tooltip"`
		Gradient    bool   `json:"gradient,omitempty"`
	} `json:"display"`
	Permissions struct {
		MinRole UserRole `json:"min_role"`
	} `json:"permissions"`
}

// AgentClaims são os claims extraídos do JWT do agente
type AgentClaims struct {
	ClientID   string  `json:"sub"`
	CNPJ       string  `json:"cnpj"`
	Empresas   []int64 `json:"empresas"`
	Plano      string  `json:"plano"`
	Type       string  `json:"type"`
}

// UserClaims são os claims extraídos do JWT do usuário PWA
type UserClaims struct {
	UserID    string   `json:"sub"`
	ClientID  string   `json:"client_id"`
	CNPJ      string   `json:"cnpj"`
	Role      UserRole `json:"role"`
	Empresas  []int64  `json:"empresas"`
}

// DescontoRequest é publicado pelo agente no tópico sga/{cnpj}/autorizacoes
// quando o PDV solicita autorização de desconto ao proprietário/gerente.
type DescontoRequest struct {
	Type                string  `json:"type"`                 // sempre "DESCONTO_REQUEST"
	NotificationID      string  `json:"notification_id"`      // UUID gerado pelo agente
	Bico                int     `json:"bico"`
	Litros              float64 `json:"litros"`
	ValorTotal          float64 `json:"valor_total"`
	DescontoSolicitado  float64 `json:"desconto_solicitado"`  // percentual ex: 5.0
	OperadorID          string  `json:"operador_id"`
	OperadorNome        string  `json:"operador_nome,omitempty"`
	Timestamp           int64   `json:"timestamp"`            // UnixMilli
}

// AuditEntry representa um registro de auditoria
type AuditEntry struct {
	Timestamp   time.Time
	UserID      string
	UserRole    UserRole
	Acao        string
	Tabela      string
	RegistroID  string
	ValorAntes  interface{}
	ValorDepois interface{}
	IPOrigem    string
	Status      string // SUCCESS | DENIED | ERROR
}
